dotcom-rendering/src/components/AdPortals.importable.tsx

```diff
@@ -158,14 +103,52 @@ export const AdPortals = () => {
 	 * Once the list of ad placeholders is available, the last render should
 	 * have populated the client-side ad slots. We can use references to those
 	 * ad slots to calculate their positions, and pass these to the native
-	 * layer so that it can insert ads.
+	 * layer so that it can insert ads. If the number of ad slots doesn't match
+	 * the number of placeholders, then the ad slot references haven't been set
+	 * up correctly and we shouldn't try to insert or update ads.
+	 *
+	 * Update Ads
+	 *
+	 * Ask the native layer to update the positions of the adverts whenever the
+	 * body element's height changes, because this typically means the ad slot
+	 * positions will have changed. To minimise the number of calls to the
+	 * native layer we:
+	 * - Check the new ad positions against the previous ones to make sure
+	 * they've changed.
+	 * - Debounce any calls to the Bridget `updateAdverts` method.
 	 */
 	useEffect(() => {
-		if (adPlaceholders.length !== 0) {
-			const newAdPositions = calculateAdPositions(adSlots.current);
-			void getCommercialClient().insertAdverts(newAdPositions);
-			adPositions.current = newAdPositions;
+		let resizeObserver: ResizeObserver | undefined = undefined;
+
+		if (
+			adPlaceholders.length !== 0 &&
+			adPlaceholders.length === adSlots.current.length
+		) {
+			adPositions.current = calculateAdPositions(adSlots.current);
+			bodyHeight.current = document.body.clientHeight;
+			void getCommercialClient().insertAdverts(adPositions.current);
+
+			resizeObserver = new ResizeObserver((entries) => {
+				if (
+					entries[0] !== undefined &&
+					entries[0].target.clientHeight !== bodyHeight.current
+				) {
+					bodyHeight.current = entries[0].target.clientHeight;
+					const newAdPositions = calculateAdPositions(
+						adSlots.current,
+					);
+
+					if (adsHaveMoved(adPositions.current, newAdPositions)) {
+						void debounceUpdateAds(newAdPositions);
+						adPositions.current = newAdPositions;
+					}
+				}
+			});
+
+			resizeObserver.observe(document.body);
 		}
+
+		return () => resizeObserver?.disconnect();
 	}, [adPlaceholders]);
```

Comment:
If we are dependent on adPlaceholders here, does that mean we're destroying and  recreating the resize observer when the adPlaceholders array changes? I guess this would only be once so maybe this is not much of a concern